File: typescript/day6_part1_2016.ts
import * as fs from 'fs';
import * as path from 'path';

// Function to read the input file
function readInputFile(filePath: string): string[] {
    const fileContents = fs.readFileSync(filePath, 'utf-8');
    return fileContents.split('\n').filter(line => line.trim() !== '');
}

// Function to find the most frequent character for each position
function findMostFrequentCharacters(lines: string[]): string {
    const numLines = lines.length;
    const numColumns = lines[0].length;
    const frequencyMap: { [key: string]: number[] } = {};

    // Initialize the frequency map
    for (let i = 0; i < numColumns; i++) {
        frequencyMap[i] = Array(26).fill(0); // Assuming only lowercase letters
    }

    // Count the occurrences of each character in each column
    for (let line of lines) {
        for (let i = 0; i < numColumns; i++) {
            const charCode = line.charCodeAt(i) - 'a'.charCodeAt(0);
            frequencyMap[i][charCode]++;
        }
    }

    // Determine the most frequent character for each column
    let result = '';
    for (let i = 0; i < numColumns; i++) {
        let maxFrequency = 0;
        let mostFrequentChar = '';
        for (let j = 0; j < 26; j++) {
            if (frequencyMap[i][j] > maxFrequency) {
                maxFrequency = frequencyMap[i][j];
                mostFrequentChar = String.fromCharCode(j + 'a'.charCodeAt(0));
            }
        }
        result += mostFrequentChar;
    }

    return result;
}

// Main function to execute the program
function main() {
    const filePath = path.join(__dirname, 'input.txt');
    const lines = readInputFile(filePath);
    const errorCorrectedMessage = findMostFrequentCharacters(lines);
    console.log(errorCorrectedMessage);
}

// Execute the main function
main();

File: fortran90/day6_part1_2016.f90
program error_corrected_message
  implicit none
  integer, parameter :: max_length = 100
  integer :: i, j, k, n
  character(len=max_length) :: line
  character(len=max_length) :: message
  integer :: counts(max_length, 26)
  character :: most_frequent

  open(unit=10, file='input.txt', status='old')

  ! Initialize counts array
  counts = 0

  ! Read input file and count occurrences of each character
  do
    read(10, '(a)', iostat=i) line
    if (i /= 0) exit
    do j = 1, len(line)
      k = ichar(line(j:j)) - ichar('a') + 1
      counts(j, k) = counts(j, k) + 1
    end do
  end do

  ! Find most frequent character for each position
  message = ''
  do j = 1, max_length
    most_frequent = ' '
    n = 0
    do k = 1, 26
      if (counts(j, k) > n) then
        n = counts(j, k)
        most_frequent = char(ichar('a') + k - 1)
      end if
    end do
    message(j:j) = most_frequent
  end do

  ! Print error-corrected message
  write (*, '(a)') trim(message)

  close(10)
end program error_corrected_message